lgc/state/PalMetadata.cpp

```diff
@@ -520,33 +543,63 @@ void PalMetadata::getVertexFetchInfo(SmallVectorImpl<VertexFetchInfo> &fetches)
-    else if (tyName == "i32")
-      ty = Type::getInt32Ty(m_pipelineState->getContext());
-    else if (tyName == "i64")
-      ty = Type::getInt64Ty(m_pipelineState->getContext());
-    else if (tyName == "f16")
-      ty = Type::getHalfTy(m_pipelineState->getContext());
-    else if (tyName == "f32")
-      ty = Type::getFloatTy(m_pipelineState->getContext());
-    else if (tyName == "f64")
-      ty = Type::getDoubleTy(m_pipelineState->getContext());
-    if (vecLength != 0)
-      ty = FixedVectorType::get(ty, vecLength);
+    Type *ty = getLlvmType(tyName);
     fetches.push_back({location, component, ty});
   }
   m_pipelineNode.erase(m_document->getNode(PipelineMetadataKey::VertexInputs));
 }
 
+// =====================================================================================================================
+// Store the color export information in PAL metadata for an exportless fragment shader with shader compilation.
+//
+// @param info : Array of ColorExportInfo structs
+void PalMetadata::addColorExportInfo(ArrayRef<ColorExportInfo> exports) {
+  // Each color export is an array containing {location, original location,type}.
+  // .colorExports is an array containing the color exports.
+  m_colorExports = m_pipelineNode[PipelineMetadataKey::ColorExports].getArray(true);
+  for (const ColorExportInfo &exportInfo : exports) {
+    msgpack::ArrayDocNode fetchNode = m_document->getArrayNode();
+    fetchNode.push_back(m_document->getNode(exportInfo.hwColorTarget));
+    fetchNode.push_back(m_document->getNode(exportInfo.location));
+    fetchNode.push_back(m_document->getNode(exportInfo.isSigned));
+    fetchNode.push_back(m_document->getNode(getTypeName(exportInfo.ty), /*copy=*/true));
+    m_colorExports.push_back(fetchNode);
+  }
+}
+
+// =====================================================================================================================
+// Get the count of color exports needed by the fragement shader.
+unsigned PalMetadata::getColorExportCount() {
+  if (m_colorExports.isEmpty())
+    return 0;
+  return m_colorExports.size();
+}
+
+// =====================================================================================================================
+// Get the color export information out of PAL metadata. Used by the linker to generate the color export shader, and to
+// finalize the PS register information. Also removes the color export information, so it does not appear in the final
+// linked ELF.
+//
+// @param [out] exports : Vector to store info of each export
+void PalMetadata::getColorExportInfo(llvm::SmallVectorImpl<ColorExportInfo> &exports) {
+  if (m_colorExports.isEmpty()) {
+    auto it = m_pipelineNode.find(m_document->getNode(PipelineMetadataKey::ColorExports));
+    if (it == m_pipelineNode.end() || !it->second.isArray())
+      return;
+    m_colorExports = it->second.getArray();
+    m_pipelineNode.erase(m_document->getNode(PipelineMetadataKey::ColorExports));
```

Comment:
Just `erase(it)`?

Reply:
Yeah, I guess I should have called it getAndEraseVertexFetchInfo.

I don't think you can erase it there, before you have even iterated it and copied the info out.

Reply:
I need to be able to call this twice to get the data.  I had assumed that the ArrayDocNode saved in m_colorExports would not have been deleted because this is being cached.  So I'll separate this into two functions.  One to get the data, the other to erase.